### PR TITLE
Add catalan_numbers.py

### PR DESCRIPTION
**Pull request by Kommandat**
### **Describe your change:**
* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### Additional Notes

Excited to make my first contribution to The Algorithms! Catalan numbers are a sequence in mathematics that represent answers to many combinatorics problems. They can be calculated using a recurrence relation and we can use tabulation (similarly to how we can for the Fibonacci sequence). I thought this best fit in the "dynamic_programming" folder given we're using tabulation to compute the sequence.

All work is my own, but I did take a look at the existing [C++ Catalan number algorithm](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/dynamic_programming/catalan_numbers.cpp) and [Python Fibonacci algorithm ](https://github.com/TheAlgorithms/Python/blob/master/dynamic_programming/fibonacci.py) for inspiration and best practices. If this is approved, I hope to contribute to Algorithms-Explanation as well.

Please let me know if you have any feedback!

Sources:
[1] https://brilliant.org/wiki/catalan-numbers/
[2] https://en.wikipedia.org/wiki/Catalan_number

### CODE REVIEW

File: dynamic_programming/catalan_numbers.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * C(n) = (1 / (n + 1)) * (2n choose n)
+
+ * Sources:
+ *  [1] https://brilliant.org/wiki/catalan-numbers/
+ *  [2] https://en.wikipedia.org/wiki/Catalan_number
+"""
+
+
+class Catalan:
+    def __init__(self, N: int) -> None:
```
[the remark this answers is not in the document]
Comment (Kommandat) [reply dynamic_programming/catalan_numbers.py:31]: I added a comment explaining N's function below this line and I don't think a more descriptive variable name is necessary

**Comment by appgurueu** [comment]
I don't like the unneeded OOP.

File: dynamic_programming/catalan_numbers.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+            # Recurrence relation: C(i) = sum(C(j).C(i-j-1)), from j = 0 to i
+            for i in range(2, upper_limit + 1):
+                for j in range(i):
+                    self.catalan_numbers[i] += (
+                        self.catalan_numbers[j] * self.catalan_numbers[i - j - 1]
+                    )
+        elif upper_limit == 0:
+            self.catalan_numbers[0] = 1
+        print(self.catalan_numbers)
```
Comment (mrmaxguns) [inline dynamic_programming/catalan_numbers.py:50]: I would avoid print statements in the algorithm itself as they produce unwanted side-effects.

File: dynamic_programming/catalan_numbers.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        >>> Catalan(5).get(6)
+        [1, 1, 2, 5, 14, 42]
+        Out of bound.
+        >>> Catalan(5).get(-1)
+        [1, 1, 2, 5, 14, 42]
+        []
+        """
+        if sequence_no is not None:
+            if sequence_no < len(self.catalan_numbers):
+                return print(self.catalan_numbers[: sequence_no + 1])
```
Comment (mrmaxguns) [inline dynamic_programming/catalan_numbers.py:66]: This should return the value directly for the same reason described above. This is where the algorithm produces its output, so the output should be directly returned.

File: dynamic_programming/catalan_numbers.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            else:
+                print("Out of bound.")
```
Comment (mrmaxguns) [inline dynamic_programming/catalan_numbers.py:68]: This is an error, so an exception should be raised here.

File: dynamic_programming/catalan_numbers.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        else:
+            print("Please specify a value")
```
Comment (mrmaxguns) [inline dynamic_programming/catalan_numbers.py:70]: Again, this should be refactored as an exception.

**Review by mrmaxguns — CHANGES_REQUESTED** [review]
Hello. Thanks for your contribution. I performed a review. If you disagree with the changes I suggested, feel free to explain your disagreement by commenting on the review.

Mostly, I suggested changes to avoid unwanted side-effects. These changes are based on the following contribution guideline:
> have minimal side effects (Ex. print(), plot(), read(), write()).

Also, when the user enters erroneous input, an exception should be raised, instead of it simply being printed (this applies only to the algorithm itself, not in the `if __name__ == "__main__"` block).
> raise Python exceptions (ValueError, etc.) on erroneous input values

**Comment by Kommandat** [comment]
> Hello. Thanks for your contribution. I performed a review. If you disagree with the changes I suggested, feel free to explain your disagreement by commenting on the review.
> 
> Mostly, I suggested changes to avoid unwanted side-effects. These changes are based on the following contribution guideline:
> 
> > have minimal side effects (Ex. print(), plot(), read(), write()).
> 
> Also, when the user enters erroneous input, an exception should be raised, instead of it simply being printed (this applies only to the algorithm itself, not in the `if __name__ == "__main__"` block).
> 
> > raise Python exceptions (ValueError, etc.) on erroneous input values

@mrmaxguns Thanks for the review! I agree with all your comments and made the changes you recommended. I requested a re-review, please take another look when you get a chance.

As an aside, I based much of my code on the [Fibonacci algorithm](https://github.com/TheAlgorithms/Python/blob/master/dynamic_programming/fibonacci.py) in the same folder thinking an existing algorithm already followed the guidelines but I was mistaken. It would nice to add some examples to the contribution guidelines that demonstrate best practices so others don't do what I did.

Once again, thanks!

**Review by mrmaxguns — APPROVED** [review]
Thank you for your contribution! Yes, indeed, some code in this repository does not follow the contribution guidelines, as they may have changed or the code was improperly reviewed. If you notice such code, feel free to open a pull request addressing the changes.

Regarding the code, it looks good. Thanks for applying the suggestions! This should be merged once tests are run and I double check the code for any issues I might not have caught now.

**Comment by mrmaxguns** [comment]
Thank you for your contribution! :100:

**Pull request merged**